File: python/src/coasty/env.py
"""Environment / configuration helpers for the Coasty cookbook.

Loads the repo-root ``.env`` (if present) exactly once via python-dotenv.
Values are only ever read from ``os.environ`` -- they are never logged.
"""

from __future__ import annotations

import os
from pathlib import Path

from dotenv import load_dotenv

DEFAULT_BASE_URL = "https://coasty.ai/v1"
SANDBOX_KEY_PREFIX = "sk-coasty-test-"

_TRUTHY = frozenset({"1", "true", "yes", "on"})

# Flipped to True after the first load attempt (tests pin it to True so the
# real repo .env is never consulted).
_dotenv_loaded = False


class MissingAPIKeyError(RuntimeError):
    """Raised when an API key is required but COASTY_API_KEY is not set."""


def _repo_root_env_file() -> Path:
    """The cookbook repo root ``.env`` (python/src/coasty -> repo root)."""
    return Path(__file__).resolve().parents[3] / ".env"


def load_repo_dotenv(*, force: bool = False) -> None:
    """Load the repo-root ``.env`` into ``os.environ`` (once, never overriding).

    Existing environment variables always win. Values are never printed.
    """
    global _dotenv_loaded
    if _dotenv_loaded and not force:
        return
    _dotenv_loaded = True
    env_file = _repo_root_env_file()
    if env_file.is_file():
        load_dotenv(env_file, override=False)


def get_api_key() -> str | None:
    """Return COASTY_API_KEY (stripped), or None when unset/empty."""
    load_repo_dotenv()
    key = os.environ.get("COASTY_API_KEY", "").strip()
    return key or None


def require_api_key() -> str:
    """Return COASTY_API_KEY or raise :class:`MissingAPIKeyError`."""
    key = get_api_key()
    if key is None:
        raise MissingAPIKeyError(
            "COASTY_API_KEY is not set. Export it or add it to the repo-root .env "
            "(use an sk-coasty-test-... sandbox key for free, unbilled testing)."
        )
    return key


def get_base_url() -> str:
    """Base URL for the API: COASTY_BASE_URL override or the default."""
    load_repo_dotenv()
    override = os.environ.get("COASTY_BASE_URL", "").strip().rstrip("/")
    return override or DEFAULT_BASE_URL


def spend_confirmed() -> bool:
    """True when COASTY_CONFIRM_SPEND opts in to billable calls (e.g. ``1``)."""
    load_repo_dotenv()
    return os.environ.get("COASTY_CONFIRM_SPEND", "").strip().lower() in _TRUTHY


def is_sandbox_key(key: str) -> bool:
    """True for sandbox keys (``sk-coasty-test-...``) which never bill."""
    return key.startswith(SANDBOX_KEY_PREFIX)

File: typescript/src/coasty/env.ts
/**
 * Environment / configuration helpers for the Coasty cookbook.
 *
 * Loads the repo-root `.env` (if present) exactly once via dotenv. Values are
 * merged into `process.env` and are NEVER logged by this module.
 */
import { existsSync } from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import dotenv from 'dotenv';

export const DEFAULT_BASE_URL = 'https://coasty.ai/v1';
export const SANDBOX_KEY_PREFIX = 'sk-coasty-test-';

let dotenvLoaded = false;

function repoRootEnvPath(): string {
  // This file lives at <repo>/typescript/src/coasty/env.ts -> repo root is 3 levels up.
  const here = path.dirname(fileURLToPath(import.meta.url));
  return path.resolve(here, '..', '..', '..', '.env');
}

/** Load the repo-root `.env` into `process.env` once. Existing vars win. */
export function loadEnvFile(): void {
  if (dotenvLoaded) return;
  dotenvLoaded = true;
  const envPath = repoRootEnvPath();
  if (existsSync(envPath)) {
    // `quiet` suppresses dotenv's injection banner; values are never printed.
    dotenv.config({ path: envPath, quiet: true });
  }
}

/** Thrown when `COASTY_API_KEY` is missing. */
export class MissingApiKeyError extends Error {
  constructor() {
    super(
      'COASTY_API_KEY is not set. Export it in your shell or add it to the repo-root .env file ' +
        '(see .env.example). Use an sk-coasty-test-* sandbox key for free, unbilled development.',
    );
    this.name = 'MissingApiKeyError';
  }
}

/**
 * The API key from `COASTY_API_KEY` (after loading `.env`).
 * @throws {MissingApiKeyError} when unset or blank.
 */
export function getApiKey(env: NodeJS.ProcessEnv = process.env): string {
  if (env === process.env) loadEnvFile();
  const key = env.COASTY_API_KEY?.trim();
  if (key === undefined || key === '') throw new MissingApiKeyError();
  return key;
}

/** Base URL from `COASTY_BASE_URL`, defaulting to the public API. Trailing slashes stripped. */
export function getBaseUrl(env: NodeJS.ProcessEnv = process.env): string {
  if (env === process.env) loadEnvFile();
  const raw = env.COASTY_BASE_URL?.trim();
  const url = raw === undefined || raw === '' ? DEFAULT_BASE_URL : raw;
  return url.replace(/\/+$/, '');
}

/** True when the user pre-confirmed spend via `COASTY_CONFIRM_SPEND=1` (or `true`). */
export function spendConfirmed(env: NodeJS.ProcessEnv = process.env): boolean {
  if (env === process.env) loadEnvFile();
  const value = env.COASTY_CONFIRM_SPEND?.trim().toLowerCase();
  return value === '1' || value === 'true';
}

/**
 * True when the given key (or `COASTY_API_KEY` if omitted) is a sandbox key
 * (`sk-coasty-test-*` — never bills).
 */
export function isSandboxKey(key?: string, env: NodeJS.ProcessEnv = process.env): boolean {
  if (key === undefined && env === process.env) loadEnvFile();
  const candidate = key ?? env.COASTY_API_KEY;
  return typeof candidate === 'string' && candidate.startsWith(SANDBOX_KEY_PREFIX);
}
